### PR TITLE
[Update] golang example for developer guide

### PR DESCRIPTION
**Pull request by Icemap**
### First-time contributors' checklist 

- [ ] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions](https://github.com/pingcap/docs/blob/master/CONTRIBUTING.md#guideline-for-choosing-the-affected-versions).

- [ ] master (the latest development version)
- [ ] v6.2 (TiDB 6.2 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/docs-cn/pull/9556
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: develop/dev-guide-sample-application-golang.md [hunk 1]
```diff
@@ -0,0 +1,899 @@
+---
+title: Build a Simple CRUD App with TiDB and Golang
+summary: Learn how to build a simple CRUD application with TiDB and Golang.
```
Comment (Oreoxmt) [inline develop/dev-guide-sample-application-golang.md:3]:
```suggestion
summary: Learn how to build a simple CRUD application with TiDB and Golang.
aliases: ['/tidb/dev/dev-guide-outdated-for-go-sql-driver-mysql','/tidb/dev/dev-guide-outdated-for-gorm']
```

File: develop/dev-guide-optimistic-and-pessimistic-transaction.md [hunk 2]
```diff
@@ -561,17 +1003,46 @@ Change it to the following to point to the optimistic transaction example.
 <mainClass>com.pingcap.txn.optimistic.TxnExample</mainClass>
 ```
 
+</div>
+
+<div label="Golang" href="optimistic-code-golang">
+
+The Golang example in the [Write a pessimistic transaction example](#write-a-pessimistic-transaction-example) section already supports optimistic transactions and can be used directly without changes.
```
[the remark this answers is not in the document]
Comment (Icemap) [reply develop/dev-guide-optimistic-and-pessimistic-transaction.md:1010]: Same

File: develop/dev-guide-optimistic-and-pessimistic-transaction.md [hunk 3]
```diff
@@ -258,17 +330,364 @@ public class TxnExample {
...
+
+func createUser(txn *util.TiDBSqlTx, id int, nickname string, balance decimal.Decimal) error {
+    _, err := txn.ExecContext(context.Background(),
+        "INSERT INTO `users` (`id`, `nickname`, `balance`) VALUES (?, ?, ?)",
+        id, nickname, balance)
+    return err
+}
+```
+
+Then write a `txn.go` with a `main` function to call `helper.go` and handle the incoming command line arguments:
```
[the remark this answers is not in the document]
Comment (Icemap) [reply develop/dev-guide-optimistic-and-pessimistic-transaction.md:587]: Same

File: develop/dev-guide-optimistic-and-pessimistic-transaction.md [hunk 4]
```diff
@@ -258,17 +330,364 @@ public class TxnExample {
 }
 ```
 
+</div>
+
+<div label="Golang" href="pessimstic-code-golang">
+
+Write a `helper.go` file that contains the required database operations:
```
[the remark this answers is not in the document]
Comment (Icemap) [reply develop/dev-guide-optimistic-and-pessimistic-transaction.md:337]: Same

File: develop/dev-guide-optimistic-and-pessimistic-transaction.md [hunk 5]
```diff
@@ -25,11 +25,83 @@ These operations must either all succeed or all fail. You must ensure that overs
...
+
+<div label="Java" href="pessimstic-concurrent-save-java">
+
 Because you use multiple threads to simulate the situation that multiple users insert data simultaneously, you need to use a connection object with safe threads. Here use Java's popular connection pool [HikariCP](https://github.com/brettwooldridge/HikariCP) for demo.
 
+</div>
+
+<div label="Golang" href="pessimstic-concurrent-save-golang">
+
+`sql.DB` in Golang is concurrency-safe, so there is no need to import a third-party package.
```
[the remark this answers is not in the document]
Comment (Icemap) [reply develop/dev-guide-optimistic-and-pessimistic-transaction.md:38]: Same

**Review by Oreoxmt — APPROVED** [review]
LGTM

File: develop/dev-guide-delete-data.md [hunk 6]
```diff
@@ -184,4 +229,65 @@ public class BatchDeleteExample
 }
 ```
 
-In each iteration, `SELECT` selects up to 1000 rows of primary key values for data in the time period from `2022-04-15 00:00:00` to `2022-04-15 00:15:00`. Then it performs bulk-delete. Note that `TimeUnit.SECONDS.sleep(1);` at the end of each loop will cause the bulk-delete operation to pause for 1 second, preventing the bulk-delete operation from consuming too many hardware resources.
\ No newline at end of file
+In each iteration, `DELETE` up to 1000 rows data in the time period from `2022-04-15 00:00:00` to `2022-04-15 00:15:00`.
```
Comment (Oreoxmt) [inline develop/dev-guide-delete-data.md:232]:
```suggestion
In each iteration, `DELETE` deletes up to 1000 rows data in the time period from `2022-04-15 00:00:00` to `2022-04-15 00:15:00`.
```

File: develop/dev-guide-delete-data.md [hunk 7]
```diff
@@ -130,6 +170,11 @@ First, you write a `SELECT` query in a loop of your application or script. Use t
 
 Suppose you find an application error within a specific time period. You need to delete all the data for the [rating](/develop/dev-guide-bookshop-schema-design.md#ratings-table) within this period, for example, from `2022-04-15 00:00:00` to `2022-04-15 00:15:00`, and more than 10,000 records are written in 15 minutes. You can perform as follows.
 
+<SimpleTab>
+<div label="Java">
+
+In Java, the bulk-delete example is like the following:
```
Comment (qiancai) [inline develop/dev-guide-delete-data.md:176]:
```suggestion
In Java, the bulk-delete example is as follows:
```

File: develop/dev-guide-delete-data.md [hunk 8]
```diff
@@ -184,4 +229,65 @@ public class BatchDeleteExample
 }
 ```
 
-In each iteration, `SELECT` selects up to 1000 rows of primary key values for data in the time period from `2022-04-15 00:00:00` to `2022-04-15 00:15:00`. Then it performs bulk-delete. Note that `TimeUnit.SECONDS.sleep(1);` at the end of each loop will cause the bulk-delete operation to pause for 1 second, preventing the bulk-delete operation from consuming too many hardware resources.
\ No newline at end of file
+In each iteration, `DELETE` deletes up to 1000 rows in the time period from `2022-04-15 00:00:00` to `2022-04-15 00:15:00`.
```
Comment (qiancai) [inline develop/dev-guide-delete-data.md:232]:
```suggestion
In each iteration, `DELETE` deletes up to 1000 rows from `2022-04-15 00:00:00` to `2022-04-15 00:15:00`.
```

File: develop/dev-guide-delete-data.md [hunk 9]
```diff
@@ -184,4 +229,65 @@ public class BatchDeleteExample
...
+
+</div>
+
+<div label="Golang">
+
+In Golang, the bulk-delete example is like the following:
```
Comment (qiancai) [inline develop/dev-guide-delete-data.md:238]:
```suggestion
In Golang, the bulk-delete example is as follows:
```

File: develop/dev-guide-optimistic-and-pessimistic-transaction.md [hunk 10]
```diff
@@ -25,11 +25,83 @@ These operations must either all succeed or all fail. You must ensure that overs
...
+
+To adapt TiDB transactions, write a toolkit [util](https://github.com/pingcap-inc/tidb-example-golang/tree/main/util)according to the following code:
```
Comment (qiancai) [inline develop/dev-guide-optimistic-and-pessimistic-transaction.md:40]:
```suggestion
To adapt TiDB transactions, write a toolkit [util](https://github.com/pingcap-inc/tidb-example-golang/tree/main/util) according to the following code:
```

File: develop/dev-guide-sample-application-golang.md [hunk 11]
```diff
@@ -0,0 +1,900 @@
...
+
+    // So player 1 has to reduce the incoming quantity to two.
+    fmt.Println("\nbuyGoods:\n    => this trade will success")
+    if err := buyGoods(db, player2.ID, player1.ID, 2, 100); err != nil {
+        panic(err)
+    }
+}
+
+func simpleExample(db *gorm.DB) {
+    // Create a player, who has a coin and a goods..
```
Comment (qiancai) [inline develop/dev-guide-sample-application-golang.md:640]:
```suggestion
    // Create a player, who has a coin and a goods.
```

File: develop/dev-guide-sample-application-golang.md [hunk 12]
```diff
@@ -0,0 +1,900 @@
...
+
+Or you can execute the following command:
+
+{{< copyable "shell-regular" >}}
+
+```shell
+mysql --host 127.0.0.1 --port 4000 -u root<sql/dbinit.sql
+```
+
+If you are using a non-local cluster or MySQL client has not been installed, connect to your cluster and run the statement in the `sql/dbinit.sql` file.
```
Comment (qiancai) [inline develop/dev-guide-sample-application-golang.md:758]:
```suggestion
If you are using a non-local cluster or MySQL client that has not been installed, connect to your cluster and run the statement in the `sql/dbinit.sql` file.
```

**Review by qiancai — APPROVED** [review]

**Pull request merged**